RFCs/0035-partiql-window-function.md

```diff
@@ -0,0 +1,1204 @@
+Next, we demonstrate how to perform operations over the produced binding tuple.
+Finally, we show that PartiQL’s window function semantic is backward compatible to SQL.
+
+## Motivation
+
+Window functions, also known as analytic OLAP functions, have been introduced as part of the SQL standard since 2003. Window functions allow users to express many useful query, such as time series analysis, ranking, moving averages, etc. Formulating such queries in plain SQL-92 can be cumbersome and inefficient, if even possible.
+
+This RFC purposes the semantics for **window partition** in PartiQL. Subsequent RFCs are expected to cover other concepts in window function such as **peer** and **window frame**.
+
+## Guide-level explanation
+
+### Out of scope
+
+* This RFC focus on explaining the semantics of PartiQL’s window functions. The PartiQL core syntax provided in this RFC is for guidance only. Finalizing grammar and syntax is out of the scope, excluding the supported SQL compatible window function. 
+* SQL’s window implicitly defines three concepts, partition, frame, and peer. As a first step, this RFC focus on partition only, the concept of frame and peer are out of scope for now, but some preliminary work has been done to make sure the other concepts can fit into the framework in a modular fashion.
+* An exhaustive mapping between SQL’s window function and PartiQL’s expression is out of the scope. In this RFC, we only define the mapping between SQL’s `lag` function and it's corresponding PartiQL expression as an example. The mapping between other functions are skipped.
+
+We assume that the readers of this RFC has basic understanding of SQL’s window function and PartiQL semantics.
+
+### Terminology
+
+1. RFC key word: MUST, MUST NOT, REQUIRED, SHALL, SHALL NOT, etc. See Appendix (1) for detail. 
+2. Windowed Table and Window: Windowed table is a defined in SQL spec 2011, section 4.15.14. A **windowed table** is a table together with one or more windows. A **window** is a transient data structure associated
+   with a table expression. A window is defined explicitly by a window definition or implicitly by an inline window specification. Implicitly defined windows have an implementation-dependent window name. A
+   **window** is used to specify window partitions and window frames, which are collections of rows used in the definition of **window functions**.
+3. Partition: To distinguish between SQL's definition of window, in this RFC, the term **partition** refers to the window partition specified for each row. See [SQL’s Windowed Table As Nested Data](#SQL’s-Windowed-Table-As-Nested-Data) for details. 
+4. Equivalent Partition: Two binding tuples $b, b' \in B^{in}_{WINDOWED}$ are in the same equivalence partition if and only if every partition expression $e_i$ evaluates to equivalent values $v_i$ (when evaluted on b) and $v_i'$ (when evaluated on $b'$). 
```

Comment:
I know we likely don't have it but let's also include a term for `PartiQL core` to state what we mean by that because it's been used in different sections of the doc.

Reply:
Make sense. The spec has a nice definition here. We can resurfacing it here and point to the spec. 

From the spec: 

PartiQL core and PartiQL syntactic sugar:  In the interest of precision and succinctness, we tier the PartiQL specification in two layers: The PartiQL core is a functional programming language with composable aspects. Three aspects of the PartiQL core syntax and semantics are characteristic of its functional orientation: Every (sub)query and every (sub) expression input and output PartiQL data. Second, each clause of a SELECT query is itself a function. Third, every (sub)query evaluates within the environment created by the database names and the variables of the enclosing queries.
Then we layer “syntactic sugar” features over the core. Commonly, syntactic sugar achieves well-known SQL syntax and semantics. Formally, every syntactic sugar feature is explained by reduction to the core.

Reply:
+1

Reply:
Addressed in the new commit.